ocaml/typing/typemod.ml

```diff
@@ -1991,20 +2007,34 @@ and transl_signature env (sg : Parsetree.signature) =
        sg
     )
 
-and transl_modtype_decl env pmtd =
+and transl_modtype_decl ~expected_modtype_decl env pmtd =
   Builtin_attributes.warning_scope pmtd.pmtd_attributes
-    (fun () -> transl_modtype_decl_aux env pmtd)
+    (fun () -> transl_modtype_decl_aux env pmtd ~expected_modtype_decl)
 
-and transl_modtype_decl_aux env
+and transl_modtype_decl_aux ~expected_modtype_decl env
```

Comment:
With the above suggested change, this should be renamed to `~context` or similar.

Reply:
This is the special case, in that this can be called in both signatures and structures. The word "expected" in the type checker has a clear meaning which doesn't include the case when it's called in signatures.